src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1125 @@
+            in the shifted basis
+
+        Integer-valued polynomial rings commute with their base ring::
+
+            sage: K = IntegerValuedPolynomialRing(QQ).S()
+            sage: a = K.gen()
+            sage: K.is_commutative()
+            True
+            sage: L = IntegerValuedPolynomialRing(K).S()
+            sage: c = L.gen()
+            sage: L.is_commutative()
+            True
+            sage: s = a * c^3; s
+            S[1]*S[1] + (-6*S[1])*S[2] + 6*S[1]*S[3]
+            sage: parent(s)
+            Integer-Valued Polynomial Ring over Integer-Valued Polynomial
+            Ring over Rational Field in the shifted basis in the shifted basis
+
+        Integer-valued polynomial rings are commutative::
+
+            sage: c^3 * a == c * a * c * c
+            True
+
+        We can also manipulate elements in the basis and
+        coerce elements from our base field::
+
+            sage: F = IntegerValuedPolynomialRing(QQ).S()
+            sage: S = F.basis()
+            sage: S[2] * S[3]
+            3*S[3] - 12*S[4] + 10*S[5]
+            sage: 1 - S[2] * S[2] / 2
+            S[0] - 1/2*S[2] + 3*S[3] - 3*S[4]
+        """
+        def __init__(self, A):
+            r"""
+            Initialize ``self``.
+
+            INPUT:
+
+            - ``A`` -- a parent with realization
+
+            EXAMPLES::
+
+                sage: F = IntegerValuedPolynomialRing(QQ).S(); F
+                Integer-Valued Polynomial Ring over Rational Field
+                in the shifted basis
+                sage: TestSuite(F).run()
+            """
+            CombinatorialFreeModule.__init__(self, A.base_ring(),
+                                             A.indices(),
+                                             category=A.Bases(),
+                                             prefix="S",
+                                             latex_prefix=r"\mathbb{S}")
+
+        def _realization_name(self) -> str:
+            r"""
+            TESTS::
+
+                sage: F = IntegerValuedPolynomialRing(QQ).S()
+                sage: F._realization_name()
+                'shifted'
+            """
+            return "shifted"
+
+        # with @cached_method ?
+        def product_on_basis(self, n1, n2):
+            r"""
+            Return the product of basis elements ``n1`` and ``n2``.
+
+            INPUT:
+
+            - ``n1``, ``n2`` -- integers
+
+            EXAMPLES::
+
+                sage: A = IntegerValuedPolynomialRing(QQ).S()
+                sage: A.product_on_basis(0, 1)
+                S[1]
+                sage: A.product_on_basis(1, 2)
+                -2*S[2] + 3*S[3]
+            """
+            i = ZZ(n1)
+            j = ZZ(n2)
+            if j < i:
+                j, i = i, j
+
+            R = self.base_ring()
+            return self._from_dict({i + j - k: R((-1)**k * i.binomial(k) * (i + j - k).binomial(i))
+                                    for k in range(i + 1)})
```

Comment:
```suggestion
            return self._from_dict({i + j - k: coeff for k in range(i + 1)
                                    if coeff := R((-1)**k * i.binomial(k) * (i + j - k).binomial(i)},
                                   remove_zeros=False)
```
I am so happy I found this (not so commonly used) syntax; it's great for these types of loops.

Reply:
I am not very fond of this walrus syntax. :)

Reply:
Fair enough; this is likely a micro-optimization.